System/RTEError.cpp

```diff
@@ -7,73 +7,403 @@
 
 #include "SDL_messagebox.h"
 
+#ifdef _WIN32
+#include "Windows.h"
+#include "DbgHelp.h"
+#include "RTEStackTrace.h"
+#endif
+
 namespace RTE {
 
 	bool RTEError::s_CurrentlyAborting = false;
+	bool RTEError::s_IgnoreAllAsserts = false;
+	std::string RTEError::s_LastIgnoredAssertDescription = "";
+	std::source_location RTEError::s_LastIgnoredAssertLocation = {};
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#ifdef _WIN32
+	/// <summary>
+	/// Custom exception handler for Windows SEH.
+	/// Unfortunately this also intercepts any C++ exceptions and turns them into SE bullshit, meaning we can't get and rethrow the current C++ exception to get what() from it.
+	/// Even if we "translate" SE exceptions to C++ exceptions it's still ass and doesn't really work, so this is what it is and it is good enough.
+	/// </summary>
+	/// <param name="exceptPtr">Struct containing information about the exception. This will be provided by the OS exception handler.</param>
+	static LONG WINAPI RTEWindowsExceptionHandler([[maybe_unused]] EXCEPTION_POINTERS *exceptPtr) {
+		// This sorta half-assedly works in x86 because exception handling is slightly different, but since the main target is x64 we can just not care about it.
+		// Something something ESP. ESP is a guitar brand.
+#ifndef TARGET_MACHINE_X86
+
+		// Returns the last Win32 error in string format. Returns an empty string if there is no error.
+		static auto getLastWinErrorAsString = []() -> std::string {
+			DWORD errorMessageID = GetLastError();
+			if (errorMessageID == 0) {
+				return "";
+			}
+			LPSTR messageBuffer = nullptr;
+			DWORD messageFlags = FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS;
+
+			// This bullshit makes the error string and returns the size because we can't know it in advance. Don't think we actually care about the size when we construct string from a buffer but whatever.
+			size_t messageSize = FormatMessage(messageFlags, nullptr, errorMessageID, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), reinterpret_cast<LPSTR>(&messageBuffer), 0, nullptr);
+			std::string message(messageBuffer, messageSize);
+			LocalFree(messageBuffer);
```

Comment:
LocalFree here is cause messageBuffer is secretly a pointer, right? And so won't be deleted if it goes out of scope?

Reply:
Seems that way. `LPSTR` is just a bullshit `char *`.
It allocates the buffer with `LocalAlloc` because `FORMAT_MESSAGE_ALLOCATE_BUFFER` flag.
Doesn't work without it.
We don't *really* need to care about freeing it here because when the handler returns it's straight into termination, but whatever.